### PR TITLE
Controller Inventory Improvements

### PR DESCRIPTION
**Pull request by stefanmielke**
This PR is to bring inventory movement improvements when using a controller.

This should include fixes for #511

Here is a list of the changes:

* Removed code to find mouse position when using controller and mouse at the same time (too many issues when moving the cursor). Now it uses the last known slot selected with the controller.
* Only allow inventory movement to valid slots (eg.: rings can only go to ring slots).
* Warp from one side of the inventory to the other.
* Adjust mouse position on slots.
* Make the use and spell button equip items like shift click does.
* Move items to belt when using the spell button (use button still consumes the item when applicable).
* Move object as a whole item, not allowing movements that are not valid.
* Increased repeat interval to decrease 'false' moves (from 100ms to 150ms).
* Last 2 slots in the inventory now move to the last belt slot when moving down.
* Changed a lot of the code to try to make it more clear, but it is still a long function.

Known issues:

* When swapping items, even if the slot is not a valid place to put it, it keeps the item in it.
* When moving out of an item larger than 1x1, it moves to the top row of that item. It was made this way to not cause issues when picking up and dropping items.

Future enhancements:

* Option to 'save' cursor slot when reopening the inventory screen. Currently it uses the same 'slot' variable as other screens (like the character screen).

Remarks:

* I just fetched from latest, so combat is broken on this branch.

### CODE REVIEW

File: Source/controls/plrctrls.cpp [hunk 1]
```diff
@@ -490,183 +490,457 @@ void AttrIncBtnSnap(AxisDirection dir)
...
+            int iv = ii;
+            if (ii <= 0) {
+                iv = -ii;
+            }
+
+            ItemStruct &item = plr[myplr].InvList[iv - 1];
+            if (!item.isEmpty()) {
+                std::pair<int, int> size = GetInvItemSize(item._iCurs + CURSOR_FIRSTITEM);
+                size.first /= INV_SLOT_SIZE_PX;
+                size.second /= INV_SLOT_SIZE_PX;
```
Comment (Iniquitatis) [inline Source/controls/plrctrls.cpp:596]: Typo (`size.second /= INV_SLOT_SIZE_PY`)?

Comment (Iniquitatis) [reply Source/controls/plrctrls.cpp:596]: Whoops, now I feel stupid. :D

File: Source/controls/plrctrls.cpp [hunk 2]
```diff
@@ -1181,6 +1456,9 @@ void PerformSpellAction()
 		else if (pcurs > CURSOR_HAND) {
 			TryIconCurs();
 			NewCursor(CURSOR_HAND);
+		} else {
+            CheckInvItem(true);
```
Comment (AJenbo) [inline Source/controls/plrctrls.cpp:1460]: make sure to run clang-format, there are a few broken code styles in this file

**Review by AJenbo — CHANGES_REQUESTED** [review]
looks good for the most part but indentation is all over the place

File: Source/controls/plrctrls.cpp (hunk 2 above)
Comment (stefanmielke) [reply Source/controls/plrctrls.cpp:1460]: I ended up setting it up too late in the process. I'll run it correctly on the parts I changed.

**Comment by stefanmielke** [comment]
> looks good for the most part but indentation is all over the place

@AJenbo, just fixed that.

**Review by AJenbo — APPROVED** [review]

**Comment by glebm** [comment]
If I open the inventory, move the cursor, then close and open the inventory, the cursor position will initially be correct but pressing a D-Pad key will move it to the wrong position.

**Comment by glebm** [comment]
To reproduce, build like this:

```bash
cmake -S. -Bbuild-kbctrl -GNinja -DCMAKE_BUILD_TYPE=Debug -DUSE_SDL1=ON -DHAS_KBCTRL=1 -DPREFILL_PLAYER_NAME=ON \ 
-DKBCTRL_BUTTON_DPAD_LEFT=SDLK_LEFT \
-DKBCTRL_BUTTON_DPAD_RIGHT=SDLK_RIGHT \
-DKBCTRL_BUTTON_DPAD_UP=SDLK_UP \
-DKBCTRL_BUTTON_DPAD_DOWN=SDLK_DOWN \
-DKBCTRL_BUTTON_X=SDLK_y \
-DKBCTRL_BUTTON_Y=SDLK_x \
-DKBCTRL_BUTTON_B=SDLK_a \
-DKBCTRL_BUTTON_A=SDLK_b \
-DKBCTRL_BUTTON_RIGHTSHOULDER=SDLK_RIGHTBRACKET \
-DKBCTRL_BUTTON_LEFTSHOULDER=SDLK_LEFTBRACKET \
-DKBCTRL_BUTTON_LEFTSTICK=SDLK_TAB \
-DKBCTRL_BUTTON_START=SDLK_RETURN \
-DKBCTRL_BUTTON_BACK=SDLK_LSHIFT
cmake --build build-kbctrl -j $(nproc)
build-kbctrl/devilutionx
```

Then, <kbd>RETURN + ]</kbd> to open the inventory, move cursor, open and close again

**Comment by stefanmielke** [comment]
> > To reproduce, build like this:
> > ```shell
> > cmake -S. -Bbuild-kbctrl -GNinja -DCMAKE_BUILD_TYPE=Debug -DUSE_SDL1=ON -DHAS_KBCTRL=1 -DPREFILL_PLAYER_NAME=ON \ 
> > -DKBCTRL_BUTTON_DPAD_LEFT=SDLK_LEFT \
> > -DKBCTRL_BUTTON_DPAD_RIGHT=SDLK_RIGHT \
> > -DKBCTRL_BUTTON_DPAD_UP=SDLK_UP \
> > -DKBCTRL_BUTTON_DPAD_DOWN=SDLK_DOWN \
> > -DKBCTRL_BUTTON_X=SDLK_y \
> > -DKBCTRL_BUTTON_Y=SDLK_x \
> > -DKBCTRL_BUTTON_B=SDLK_a \
> > -DKBCTRL_BUTTON_A=SDLK_b \
> > -DKBCTRL_BUTTON_RIGHTSHOULDER=SDLK_RIGHTBRACKET \
> > -DKBCTRL_BUTTON_LEFTSHOULDER=SDLK_LEFTBRACKET \
> > -DKBCTRL_BUTTON_LEFTSTICK=SDLK_TAB \
> > -DKBCTRL_BUTTON_START=SDLK_RETURN \
> > -DKBCTRL_BUTTON_BACK=SDLK_LSHIFT
> > cmake --build build-kbctrl -j $(nproc)
> > build-kbctrl/devilutionx
> > ```
> > 
> > 
> >   
> >     
> >     
> >   
> > 
> > 
> > Then, RETURN + ] to open the inventory, move cursor with arrow keys, then open and close again
> 
> Yep! Able to reproduce. Interestingly it only happens when emulating the gamepad, not with an actual gamepad. And I did reproduce with an SDL2 build :)

Scratch that, it happens only when not using the button to open the inventory directly (right trigger). I'll take a look at why they are different methods.

**Comment by glebm** [comment]
@stefanmielke another issue, pressing A can place a bow into the inventory but not into a weapon slot

To reproduce, start with a rogue, click on the bow, then try to place it back into the slot

**Comment by StephenCWills** [comment]
> @stefanmielke another issue, pressing A can place a bow into the inventory but not into a weapon slot
> 
> To reproduce, start with a rogue, click on the bow, then try to place it back into the slot

Actually, I noticed the same thing with swords and shields on my Warrior today. So it's not just limited to bows.

**Comment by stefanmielke** [comment]
> > @stefanmielke another issue, pressing A can place a bow into the inventory but not into a weapon slot
> > To reproduce, start with a rogue, click on the bow, then try to place it back into the slot
> 
> Actually, I noticed the same thing with swords and shields on my Warrior today. So it's not just limited to bows.

@glebm @StephenCWills 

I was able to reproduce even with a mouse, but it's way more noticeable with a controller on 640x480 (other resolutions seem to work normally even with a controller).

The problem is when trying to equip an item when the mouse is in the exact center of the tile.

**Pull request merged**